utils/src/blockuntilzero.rs

```diff
@@ -0,0 +1,110 @@
+use atomic_traits::{Atomic, NumOps};
+use num_traits::{One, Zero};
+use std::{
+    sync::{atomic::Ordering, Arc},
+    time::Duration,
+};
+
+use crate::counttracker::CountTracker;
+
+pub struct BlockUntilZero<T>
+where
+    T: Atomic + NumOps,
+    <T as Atomic>::Type: One,
+    <T as Atomic>::Type: Zero,
+    <T as Atomic>::Type: Ord,
+{
+    value: Arc<T>,
+}
+
+impl<T> BlockUntilZero<T>
+where
+    T: Atomic + NumOps,
+    <T as Atomic>::Type: One,
+    <T as Atomic>::Type: Zero,
+    <T as Atomic>::Type: Ord,
+{
+    pub fn new() -> Self {
+        Self {
+            value: Arc::new(<T as Atomic>::new(<T as Atomic>::Type::zero())),
+        }
+    }
+
+    pub fn wait_for_zero(&self) {
+        while self.value.load(Ordering::Acquire) > <T as Atomic>::Type::zero() {
+            std::thread::yield_now();
+            std::thread::sleep(Duration::from_millis(10));
+        }
+    }
+
+    pub fn value(&self) -> <T as Atomic>::Type {
+        self.value.load(Ordering::Acquire)
+    }
+
+    #[must_use = "CountTracker is useless without holding its object"]
+    pub fn count_one(&self) -> CountTracker<T> {
+        CountTracker::new(Arc::clone(&self.value))
+    }
+}
+
+impl<T> Default for BlockUntilZero<T>
+where
+    T: Atomic + NumOps,
+    <T as Atomic>::Type: One,
+    <T as Atomic>::Type: Zero,
+    <T as Atomic>::Type: Ord,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Drop for BlockUntilZero<T>
+where
+    T: Atomic + NumOps,
+    <T as Atomic>::Type: One,
+    <T as Atomic>::Type: Zero,
+    <T as Atomic>::Type: Ord,
+{
+    fn drop(&mut self) {
+        self.wait_for_zero()
```

Comment:
I don't think this is a good idea. I've already raised my concern regarding us mixing sync and async code elsewhere so we don't need to have that discussion here but having a "hidden" synchronous blocking operation at the end of scope while non-trivial part of our blockchain is async sounds very suspicious to me. If we are going to go this route of both sync and async, I'd much rather see this wait being as explicit as possible.

Reply:
I'll see if I can wrap it even further.

About combining sync and async... quite frankly I don't know what to do. Since we started with the subsystem thing, I'm not happy that consensus will get into that soup. Like I told Lukas last week, there are certain guarantees and that's all I care about. I've never heard anyone state in C++ "mixing async and sync code" except for worries of deadlock. Quite frankly I'm failing to understand the issue.

Let me see if I can wrap it even further to make it self-contained and serve events. Though I can't promise a solution for the async/sync thing.

Reply:
Right now I think the complexity is manageable but IMO the subsystem wrapper should act as the bridge, if at all possible. I think when we start adding more complicated features like maybe the free market for gas or programmable pools the complexity will grow and at that point having one central place for the bridge seems like a good idea to me. Like for example right now in P2P I need to convert the synchronous event into an async event so it can be polled by `tokio::select!`. If this was handled by the subsystem wrapper through some async interface, that would be perfect

Reply:
I'm still undecided what the best generic solution for events should be. But I think this should do for a good start with a complete solution until the time for that discussion comes.